_posts/2015-11-21-happy-trees.md

```diff
@@ -0,0 +1,167 @@
+only happifies to itself, and nothing else will happify to it (sad `000000`).
+The second circle represents `000001`, and all of its runty children. And
+finally the melancoil, comprised of: `00000D`, `0000A9`, `0000B5`, `000092`,
+`000055`, and `000032`.
+
+The final thing you'll notice (or maybe it was the first, since it's really
+obvious) is that it's very blue. Non-blue colors are really only represented as
+leaves on their trees and don't ever really have any children of their own, so
+the blue and black sections take up vastly more space.
+
+This makes sense. The number which should generate the largest happification
+result, `FFFFFF`, only results in `000546`, which is primarily blue. So in effect
+all colors happify to some shade of blue.
+
+This might have been it, technically this is the happification tree and the
+melancoil of 6 digit hexadecimal numbers represented as colors. But it's also
+boring, and I wanted to do better.
+
+## Second attempt
+
+The root of the problem is that the definition of "happification" I used
+resulted in not diverse enough results. I wanted something which would give me
+numbers where any of the digits could be anything. Something more random.
+
+I considered using a hash instead, like md5, but that has its own problems.
+There's no gaurantee that any number would actually reach `000001`, which isn't
+required but it's a nice feature that I wanted. It also would be unlikely that
+there would be any melancoils that weren't absolutely gigantic.
+
+I ended up redefining what it meant to happify a hexadecimal number. Instead of
+adding all the digits up, I first split up the red, green, and blue digits into
+their own numbers, happified those numbers, and finally reassembled the results
+back into a single number. For example:
+
+```
+FF5500
+FF, 55, 00
+F*F + F*F, 5*5 + 5*5, 0*0 + 0*0
+1C2, 32, 00
+C23200
+```
+
+I drop that 1 on the `1C2`, because it has no place in this system. Sorry 1.
+
+Simply replacing that function resulted in this image (click for 5000x5000) version:
+
+[![Result of second attempt](/imgs/happy-tree/happy-tree-colors-small.png)](/imgs/happy-tree/happy-tree-colors.png)
+
+The first thing you notice is that it's so colorful! So that goal was achieved.
+
+The second thing you notice is that there's *significantly* more melancoils.
+Hundreds, even. I'll list out a couple of them here:
+
+```
+00000D 0000A9 0000B5 000092 000055 000032
+000D0D 00A9A9 00B5B5 009292 005555 003232
+0D0D0D A9A9A9 B5B5B5 929292 555555 323232
+0D0D32 A9A90D B5B5A9 9292B5 555592 323255
+...
+```
+
+And so on. Those numbers are actually the same numbers from the
+melancoil in the first attempt. This make sense, since the happification
+algorithm didn't really change, just how we applied it, so the melancoil numbers
+didn't change, just where they appear. And they appear *a lot*. Every
+conceivable combination of those three two-digit places across 6 numbers.
```

Comment:
Yeah I can try and do this whole section better.

Each line represents its own coil. But really all the coils are just different iterations of the original coil from the first attempt. If we break it down into 4 digits it's easier to see:

```
000D -> 00A9 -> 00B5 -> ...
0D0D -> A9A9 -> B5B5 -> ...
0D32 -> A90D -> B5A9 -> ...
...
```

Does that make it easier to see what's happening? I haven't come up with a great way to explain it =/ On the first line is the coil from the first attempt. The second line is that same coil, but with another coil happening in the first two digits. The third line is the same thing, but the coil in the first two digits is offset by one step in the coil.